pkg/controller/master/node/promote_controller.go

```diff
@@ -0,0 +1,488 @@
+	}
+
+	return job, nil
+}
+
+func (h *PromoteHandler) promote(node *corev1.Node) (*corev1.Node, error) {
+	// first, mark node into promote status
+	startedNode, err := h.setPromoteStart(node)
+	if err != nil {
+		return nil, err
+	}
+
+	// then, create a promote job on the node
+	if _, err := h.createPromoteJob(node); err != nil {
+		return nil, err
+	}
+
+	return startedNode, nil
+}
+
+func (h *PromoteHandler) logPromoteEvent(node *corev1.Node, status string) {
+	preStatus := node.Annotations[HarvesterPromoteStatusAnnotationKey]
+	eventType := corev1.EventTypeNormal
+	switch status {
+	case PromoteStatusUnknown, PromoteStatusFailed:
+		eventType = corev1.EventTypeWarning
+	}
+	nodeReference := &corev1.ObjectReference{
+		Name: node.Name,
+		UID:  types.UID(node.Name),
+		Kind: "Node",
+	}
+	h.recorder.Event(nodeReference, eventType,
+		fmt.Sprintf("NodePromote%s", strings.Title(status)),
+		fmt.Sprintf("Node %s promote status change: %s => %s", node.Name, preStatus, status))
+}
+
+// setPromoteStart set node unschedulable and set promote status running.
+func (h *PromoteHandler) setPromoteStart(node *corev1.Node) (*corev1.Node, error) {
+	if node.Annotations[HarvesterPromoteStatusAnnotationKey] == PromoteStatusRunning {
+		return node, nil
+	}
+	h.logPromoteEvent(node, PromoteStatusRunning)
+	toUpdate := node.DeepCopy()
+	toUpdate.Labels[KubeNodeSVCEnableLBLabelKey] = "true"
+	toUpdate.Annotations[HarvesterPromoteStatusAnnotationKey] = PromoteStatusRunning
+	toUpdate.Spec.Unschedulable = true
+	return h.nodes.Update(toUpdate)
+}
+
+// setPromoteResult set node schedulable and update promote status if the promote is successful
+func (h *PromoteHandler) setPromoteResult(job *batchv1.Job, node *corev1.Node, status string) (*batchv1.Job, error) {
+	if node.Annotations[HarvesterPromoteStatusAnnotationKey] == status {
+		return job, nil
+	}
+	h.logPromoteEvent(node, status)
+	toUpdate := node.DeepCopy()
+	toUpdate.Annotations[HarvesterPromoteStatusAnnotationKey] = status
+	if status == PromoteStatusComplete {
+		toUpdate.Spec.Unschedulable = false
+	}
+	_, err := h.nodes.Update(toUpdate)
+	return job, err
+}
+
+type SortByAgeNodeList []*corev1.Node
+
+func (s SortByAgeNodeList) Len() int      { return len(s) }
+func (s SortByAgeNodeList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s SortByAgeNodeList) Less(i, j int) bool {
+	// reverse order
+	return !s[i].CreationTimestamp.Before(&s[j].CreationTimestamp)
+}
+
+// selectPromoteNodeList return a list contain nodes need to be promoted
+// If the cluster doesn't need to be promoted, return nil
+func (h *PromoteHandler) selectPromoteNodeList(nodeList []*corev1.Node) ([]*corev1.Node, error) {
+	masterNumber := 0
+	canBePromoteNodeList := make([]*corev1.Node, 0, len(nodeList))
+	for _, node := range nodeList {
+		if h.shouldBeMaster(node) {
+			masterNumber++
+		} else if h.isHarvesterNode(node) && h.isHealthyNode(node) {
+			canBePromoteNodeList = append(canBePromoteNodeList, node)
+		}
```

Comment:
this is not thread-safe to count the promote node,  for example when adding both the 3rd and 4th node simultaneously via PXE, none of the existing worker nodes has been marked as `promoted-node` yet, and according to the current reconcile logic it will promote 2 groups of the master node(2nd and 3rd, 3rd and 4th) to be the new master node, as a result, we will get a total number of 4 master node.